File: tools/semantic_ir/declarations.py
"""STUNIR Semantic IR Declaration Nodes.

DO-178C Level A Compliant
Declaration node definitions.
"""

from typing import List, Optional, Union
from pydantic import BaseModel, Field, conint

from .ir_types import (
    NodeID, IRName, IRNodeKind,
    StorageClass, MutabilityKind,
    VisibilityKind, InlineHint
)
from .nodes import IRNodeBase, TypeReference


class DeclarationNode(IRNodeBase):
    """Base declaration node."""
    name: IRName
    visibility: VisibilityKind = VisibilityKind.PUBLIC


class Parameter(BaseModel):
    """Function parameter."""
    name: IRName
    param_type: TypeReference
    default: Optional[Union[NodeID, "ExpressionNode"]] = None


class FunctionDecl(DeclarationNode):
    """Function declaration."""
    kind: IRNodeKind = IRNodeKind.FUNCTION_DECL
    return_type: TypeReference
    parameters: List[Parameter] = Field(default_factory=list)
    body: Optional[Union[NodeID, "StatementNode"]] = None
    inline: InlineHint = InlineHint.NONE
    is_pure: bool = False
    stack_usage: conint(ge=0) = 0
    priority: int = 0
    interrupt_vector: conint(ge=0) = 0  # 0 means not an interrupt handler
    entry_point: bool = False


class TypeDecl(DeclarationNode):
    """Type declaration."""
    kind: IRNodeKind = IRNodeKind.TYPE_DECL
    type_definition: TypeReference


class ConstDecl(DeclarationNode):
    """Constant declaration."""
    kind: IRNodeKind = IRNodeKind.CONST_DECL
    const_type: Optional[TypeReference] = None
    value: Union[NodeID, "ExpressionNode"]
    compile_time: bool = True


class VarDecl(DeclarationNode):
    """Variable declaration."""
    kind: IRNodeKind = IRNodeKind.VAR_DECL
    var_type: Optional[TypeReference] = None
    initializer: Optional[Union[NodeID, "ExpressionNode"]] = None
    storage: StorageClass = StorageClass.AUTO
    mutability: MutabilityKind = MutabilityKind.MUTABLE
    alignment: conint(ge=1) = 1
    is_volatile: bool = False


# Enable forward references
from .expressions import ExpressionNode  # noqa: E402
from .statements import StatementNode  # noqa: E402

Parameter.model_rebuild()
FunctionDecl.model_rebuild()
ConstDecl.model_rebuild()
VarDecl.model_rebuild()

File: tools/rust/semantic_ir/src/declarations.rs
//! STUNIR Semantic IR Declaration Nodes
//!
//! Declaration node definitions.

use serde::{Deserialize, Serialize};

use crate::types::*;
use crate::nodes::*;

/// Declaration node
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct DeclarationNode {
    #[serde(flatten)]
    pub base: IRNodeBase,
    pub name: IRName,
    #[serde(default)]
    pub visibility: VisibilityKind,
}

impl Default for VisibilityKind {
    fn default() -> Self { VisibilityKind::Public }
}

/// Function parameter
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Parameter {
    pub name: IRName,
    #[serde(rename = "type")]
    pub param_type: TypeReference,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub default: Option<NodeID>,
}

/// Function declaration
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct FunctionDecl {
    #[serde(flatten)]
    pub base: DeclarationNode,
    pub return_type: TypeReference,
    #[serde(default)]
    pub parameters: Vec<Parameter>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub body: Option<NodeID>,
    #[serde(default)]
    pub inline: InlineHint,
    #[serde(default)]
    pub is_pure: bool,
    #[serde(default)]
    pub stack_usage: u32,
    #[serde(default)]
    pub priority: i32,
    #[serde(default)]
    pub interrupt_vector: u32,
    #[serde(default)]
    pub entry_point: bool,
}

impl Default for InlineHint {
    fn default() -> Self { InlineHint::None }
}

/// Type declaration
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct TypeDecl {
    #[serde(flatten)]
    pub base: DeclarationNode,
    pub type_definition: TypeReference,
}

/// Constant declaration
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ConstDecl {
    #[serde(flatten)]
    pub base: DeclarationNode,
    #[serde(skip_serializing_if = "Option::is_none", rename = "const_type")]
    pub const_type: Option<TypeReference>,
    pub value: NodeID,
    #[serde(default = "default_true")]
    pub compile_time: bool,
}

fn default_true() -> bool { true }

/// Variable declaration
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct VarDecl {
    #[serde(flatten)]
    pub base: DeclarationNode,
    #[serde(skip_serializing_if = "Option::is_none", rename = "var_type")]
    pub var_type: Option<TypeReference>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub initializer: Option<NodeID>,
    #[serde(default)]
    pub storage: StorageClass,
    #[serde(default)]
    pub mutability: MutabilityKind,
    #[serde(default = "default_alignment")]
    pub alignment: u32,
    #[serde(default)]
    pub is_volatile: bool,
}

fn default_alignment() -> u32 { 1 }
